<div style="display:block">
    <div style="width: 20%; display: inline-block; text-align: left;">
        <img src="http://upload.wikimedia.org/wikipedia/en/0/0c/Mu_Sigma_Logo.jpg" style="height:75px; margin-left:0px" />
    </div>
    <div style="width: 59%; display: inline-block">
        <h1  style="text-align: center">ML using Python: Session 1</h1>
        <div style="width: 100%; text-align: center; display: inline-block;"><i>Author:</i> <strong>Srishti Saha</strong> </div>
    </div>
    <div style="width: 20%; text-align: right; display: inline-block;">
        <div style="width: 100%; text-align: left; display: inline-block;">
            <i>Created: </i>
            <time datetime="2014-03-10" pubdate>9th February 2019</time>
        </div>
    </div>
</div>

# Topics Covered:


## 1. Sampling
find link here: [Sampling](#Sampling)
## 2. Basics of Hypothesis Testing
find link here: [Basics of Hypothesis Testing](#Basics-of-Hypothesis-Testing)
## 3. Hypothesis Testing
find link here: [Hypothesis Testing](#Hypothesis-Testing)

In [105]:
import pandas as pd
import numpy as np

## Sampling

A sampling method is a procedure for selecting sample members from a population. This is often done to reduce the size of data for an analysis depending on the ask.

References:
1. Different types of sampling: https://www.healthknowledge.org.uk/public-health-textbook/research-methods/1a-epidemiology/methods-of-sampling-population

In [87]:
# load a sample dataset
from sklearn.datasets import load_boston

#using the Boston pricing dataset
boston = load_boston()
df = pd.DataFrame(boston.data)
df.columns = boston.feature_names
print(df.head())

      CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33  


### Random Sampling


In this case each individual is chosen entirely by chance and each member of the population has an equal chance, or probability, of being selected. One way of obtaining a random sample is to give each individual in a population a number, and then use a table of random numbers to decide which individuals to include.1 For example, if you have a sampling frame of 1000 individuals, labelled 0 to 999, use groups of three digits from the random number table to pick your sample. So, if the first three numbers from the random number table were 094, select the individual labelled “94”, and so on.

In [88]:
# random sampling with number of rows
df.sample(n=3)

# random sampling with fraction
df.sample(frac=0.25) #here 25% of rows are selected

# for non overlapping two sets-- use random state
subset1 = df.sample(frac=0.25, random_state=99)
# this code tries to find that train = 75% and test = 25%
subset2 = df.loc[~df.index.isin(subset1.index), :]
print(subset1.shape)
print(subset2.shape)

(126, 13)
(380, 13)


### Stratified Sampling

In this method, the population is first divided into subgroups (or strata) who all share a similar characteristic. It is used when we might reasonably expect the measurement of interest to vary between the different subgroups, and we want to ensure representation from all the subgroups. For example, in a study of stroke outcomes, we may stratify the population by sex, to ensure equal representation of men and women. The study sample is then obtained by taking equal sample sizes from each stratum. In stratified sampling, it may also be appropriate to choose non-equal sample sizes from each stratum. For example, in a study of the health outcomes of nursing staff in a county, if there are three hospitals each with different numbers of nursing staff (hospital A has 500 nurses, hospital B has 1000 and hospital C has 2000), then it would be appropriate to choose the sample numbers from each hospital proportionally (e.g. 10 from hospital A, 20 from hospital B and 40 from hospital C). This ensures a more realistic and accurate estimation of the health outcomes of nurses across the county, whereas simple random sampling would over-represent nurses from hospitals A and B. The fact that the sample was stratified should be taken into account at the analysis stage.

In [89]:
from sklearn import model_selection

In [90]:
# choosing the column RAD for stratified sampling due to finite number of values
print(df.groupby(['RAD']).size()) #gets the distributions

df_subset1, df_subset2 = model_selection.train_test_split(df, test_size=0.25, stratify=df['RAD'])
print(df_subset1.shape)
print(df_subset2.shape)
print(df_subset1.groupby(['RAD']).size())
print(df_subset2.groupby(['RAD']).size())

RAD
1.0      20
2.0      24
3.0      38
4.0     110
5.0     115
6.0      26
7.0      17
8.0      24
24.0    132
dtype: int64
(379, 13)
(127, 13)
RAD
1.0     15
2.0     18
3.0     28
4.0     82
5.0     86
6.0     20
7.0     13
8.0     18
24.0    99
dtype: int64
RAD
1.0      5
2.0      6
3.0     10
4.0     28
5.0     29
6.0      6
7.0      4
8.0      6
24.0    33
dtype: int64


There are two more techniques for probability based sampling techniques:
1. **Systematic sampling**: Individuals are selected at regular intervals from the sampling frame. The intervals are chosen to ensure an adequate sample size. If you need a sample size n from a population of size x, you should select every x/nth individual for the sample.  For example, if you wanted a sample size of 100 from a population of 1000, select every 1000/100 = 10th member of the sampling frame.

*Kindly note: Systematic sampling is often more convenient than simple random sampling, and it is easy to administer. However, it may also lead to bias, for example if there are underlying patterns in the order of the individuals in the sampling frame, such that the sampling technique coincides with the periodicity of the underlying pattern.*

2. **Clustered Sampling**: In a clustered sample, subgroups of the population are used as the sampling unit, rather than individuals. The population is divided into subgroups, known as clusters, which are randomly selected to be included in the study. Clusters are usually already defined, for example individual GP practices or towns could be identified as clusters. In single-stage cluster sampling, all members of the chosen clusters are then included in the study. In two-stage cluster sampling, a selection of individuals from each cluster is then randomly selected for inclusion. In two-stage cluster sampling, a selection of individuals from each cluster is then randomly selected for inclusion.

*Kindly note:  Typically one would prefer cluster sampling if it is hard or expensive to visit each group/stratum as is required in stratified random sampling.*

## Basics of Hypothesis Testing

In hypothesis testing, we have to define a null hypothesis (H0) and a corresponding alternative hypothesis (Ha).

1. How do we choose between H0 and Ha? The standard procedure is to assume H0 is true - just as we presume innocent until proven guilty. Using probability theory, we try to determine whether there is sufficient evidence to declare H0 false.
2. We reject H0 only when the chance is small that H0 is true. Since our decisions
are based on probability rather than certainty, we can make errors.
3. Type I error - We reject the null hypothesis when the null is true. The probability of Type I error = α. This means, *α = Probability of Type I error = P(rejecting H0 | H0 is true)*

*Please Note: Typical values chosen for α (alpha) are .05 or .01. So, for example, if α = .05, there is a 5% chance that, when the null hypothesis is true, we will erroneously reject it.*
4. Type II error - we accept the null hypothesis when it is not true. Probability of Type II error = ß (beta). Also means, *ß = Probability of Type II error = P(accepting H0 | H0 is false)*

![Type 1 and Type 2 errors](https://wiki.mu-sigma.com/muwiki/images/8/8a/Type12_error.png)

**example of type I and type II error:**

H0: µ = 100
Ha: µ <> 100 

Suppose µ (mean) really does equal 100. But, suppose the researcher accepts HA instead. A type I error has occurred.
Or, suppose µ = 105 : but the researcher accepts H0. A type II error has occurred.

**Some important points**

1. α and ß are not independent of each other - as one increases, the other decreases. However, increases in N (sample size) cause both to decrease, since sampling error is reduced.

2. Normally, we primarily focus on Type I error. But, you should be aware that Type II error is also important. A small sample size, for example, might lead to frequent Type II errors, i.e. it could be that your (alternative) hypotheses are right, but because your sample is so small, you fail to reject the null even though you should. 

Reference:
1. About Hypothesis testing and errors: https://www3.nd.edu/~rwilliam/stats1/x24.pdf

## Hypothesis Testing

### One-tailed and Two-tailed tests

The _significance level_, also denoted as alpha or α, is the probability of rejecting the null hypothesis when it is true. For example, a significance level of 0.05 indicates a 5% risk of concluding that a difference exists when there is no actual difference.

**What is a two-tailed test?**
![Two Tailed Tests](https://wiki.mu-sigma.com/muwiki/images/1/11/Twotailed.png)
First let’s start with the meaning of a two-tailed test.  If you are using a significance level of 0.05, a two-tailed test allots half of your alpha to testing the statistical significance in one direction and half of your alpha to testing statistical significance in the other direction.  This means that .025 is in each tail of the distribution of your test statistic. When using a two-tailed test, regardless of the direction of the relationship you hypothesize, you are testing for the possibility of the relationship in both directions.  For example, we may wish to compare the mean of a sample to a given value x using a t-test.  Our null hypothesis is that the mean is equal to x. A two-tailed test will test both if the mean is significantly greater than x and if the mean significantly less than x. The mean is considered significantly different from x if the test statistic is in the top 2.5% or bottom 2.5% of its probability distribution, resulting in a p-value less than 0.05.  

**What is a one-tailed test?**
![One Tailed Tests](https://wiki.mu-sigma.com/muwiki/images/b/bb/Onetailed.png)
If you are using a significance level of .05, a one-tailed test allots all of your alpha to testing the statistical significance in the one direction of interest.  This means that .05 is in one tail of the distribution of your test statistic. When using a one-tailed test, you are testing for the possibility of the relationship in one direction and completely disregarding the possibility of a relationship in the other direction.  Let’s return to our example comparing the mean of a sample to a given value x using a t-test.  Our null hypothesis is that the mean is equal to x. A one-tailed test will test either if the mean is significantly greater than x or if the mean is significantly less than x, but not both. Then, depending on the chosen tail, the mean is significantly greater than or less than x if the test statistic is in the top 5% of its probability distribution or bottom 5% of its probability distribution, resulting in a p-value less than 0.05.  The one-tailed test provides more power to detect an effect in one direction by not testing the effect in the other direction. 

**Should you choose one tailed test?**
Although one tailed tests have more power in detecting the effect, one should consider the consequences of missing an effect in the other direction. 
If you consider the consequences of missing an effect in the untested direction and conclude that they are negligible and in no way irresponsible or unethical, then you can proceed with a one-tailed test. For example, imagine again that you have developed a new drug. It is cheaper than the existing drug and, you believe, no less effective.  In testing this drug, you are only interested in testing if it less effective than the existing drug.  You do not care if it is significantly more effective.  You only wish to show that it is not less effective. In this scenario, a one-tailed test would be appropriate. 


For more content refer to this: https://stats.idre.ucla.edu/other/mult-pkg/faq/general/faq-what-are-the-differences-between-one-tailed-and-two-tailed-tests/

References:
1. Hypothesis Testing: https://www.analyticsvidhya.com/blog/2015/09/hypothesis-testing-explained/

### Z test

Z value is a measure of standard deviation i.e. how many standard deviation away from mean is the observed value. For example, the value of z value = +1.8 can be interpreted as the observed value is +1.8 standard deviations away from the mean.
In this test, we check where the mean of an entire sample of observations falls with regards to the mean of the sampling distribution.

**Basic Assumption: Z-test is a statistical test where normal distribution is applied and is basically used for dealing with problems relating to large samples when n ≥ 30.**

A z-score can be calculated from the following formula:

_z = (X - μ) / σ_
where z is the z-score, X is the value of the element, μ is the population mean, and σ is the standard deviation.

**How to read z-scores?**

1. A z-score less than 0 represents an element less than the mean.

2. A z-score equal to 0 represents an element equal to the mean.

3. A z-score equal to 1 represents an element that is 1 standard deviation greater than the mean; a z-score equal to 2, 2 standard deviations greater than the mean; etc.

_1.In a two-tailed z-test, if z-score is greater than 1.96 or lower than -1.96 (i.e 0.025 of the total area under the curve), we reject the null hypothesis._

_2.In a one-tailed upper test, if z-score is greater than 1.645 (i.e 0.05 of the total area under the curve), we reject the null hypothesis._

_3.In a one-tailed lower test, if z-score is lesser than -1.645 (i.e 0.05 of the total area under the curve), we reject the null hypothesis._ 

Reading Material:
1. When to use Z-test: https://explorable.com/z-test
2. Steps in Z-test: http://jukebox.esc13.net/untdeveloper/RM/Stats_Module_4/mobile_pages/Stats_Module_48.html

In [91]:
import scipy.stats as st
#import scipy.special as sp

In [92]:
n = 100  # total number of samples
h = 60  # number of heads
q = .5  # null-hypothesis

xbar = float(h) / n #xbar is the estimated average of the distribution
z = (xbar - q) * np.sqrt(n / (q * (1 - q))) #z-score

# you cam also compute the p-value from here
pval = 2 * (1 - st.norm.cdf(z))

print('z-score is:', z)
print('p-value is:',pval)

z-score is: 1.9999999999999996
p-value is: 0.04550026389635842


If the p-value is less than 0.05, so we reject the null hypothesis. This is equivalent to a z-score greater than 1.96 for two-tailed tests (or less than -1.96) or 1.645 for one-tailed tests.

In [93]:
# Using pre-built package


# Demonstrating significant differences between a
# vector of measurements and a single value
# Using the statsmodels package for doing test
# Using numpy to generate some fake data

from statsmodels.stats import weightstats as stests

data=np.random.normal(loc=3.4,scale=0.1,size=50)
mean_val=3.5

In [94]:
print(np.mean(data))
data

3.3995309243434235


array([3.46688617, 3.34824285, 3.4701035 , 3.34296253, 3.33732259,
       3.61711696, 3.26877826, 3.40741498, 3.25690456, 3.27580067,
       3.27651429, 3.31794618, 3.52201481, 3.33544749, 3.67260567,
       3.32255301, 3.62123748, 3.34813981, 3.46011782, 3.43284188,
       3.31451055, 3.1463677 , 3.50018329, 3.3322575 , 3.60210048,
       3.33272494, 3.57944554, 3.48798935, 3.32731935, 3.32970678,
       3.31734386, 3.20751351, 3.54224971, 3.39977574, 3.4634891 ,
       3.39989585, 3.39758015, 3.42157339, 3.40598136, 3.40711931,
       3.37500965, 3.23087303, 3.38461542, 3.33210847, 3.41069967,
       3.6178106 , 3.44418453, 3.42810676, 3.36374396, 3.37331513])

In [95]:
# Assuming data are normally distributed, we can do z-test
test_statistic, p_value =stests.ztest(data,value=mean_val)
print("p-value is: "+str(p_value))
print("Z-test statistic is " + str(test_statistic))

p-value is: 4.2412250540942237e-10
Z-test statistic is -6.24488114209153


Here the default value of the argument 'alternative' has been taken into account. i.e. this is a two-tailed hypothesis. In this case, our alternative hypothesis was of the form:
**mean != man_val**

We can also explore the one-tailed hypotheses.
Let us set the the null hypothesis as: 
**mean < mean_value**.
Then, alternative hypothesis is mean > mean_value.
We choose the value of the argument 'alternative' as _'larger'_

In the scenario, you create a null hypothesis like mean > mean_value, choose _'smaller'_

In [96]:
test_statistic2, p_value2 = stests.ztest(data,value=mean_val, 
                                         alternative='larger')

print("Z-test statistic is " + str(test_statistic2))
print("p-value is " + str(p_value2))

Z-test statistic is -6.24488114209153
p-value is 0.9999999997879387


References:
1. http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/BS704_HypothesisTest-Means-Proportions/BS704_HypothesisTest-Means-Proportions3.html
2. https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html
3. https://www.analyticsvidhya.com/blog/2015/09/hypothesis-testing-explained/

### T Test

The t test (also called Student’s T Test) compares two averages (means) and tells you if they are different from each other. The t test also tells you how significant the differences are; In other words it lets you know if those differences could have happened by chance.

**Basic Assumptions:**
**1. Observations in each sample are independent and identically distributed (iid).**
**2. Observations in each sample are normally distributed.**
**3. Observations in each sample have the same variance.**

The t score is a ratio between the difference between two groups and the difference within the groups. 

**How to read a t-score?**

The larger the t score, the more difference there is between groups. The smaller the t score, the more similarity there is between groups. A t-score of 3 means that the groups are three times as different from each other as they are within each other. When you run a t test, the bigger the t-value, the more likely it is that the results are repeatable.

1. A large t-score tells you that the groups are different.
2. A small t-score tells you that the groups are similar.

Just like the z-score, every value of t-score corresponds to a p-value (probability that the results from your sample data occurred by chance). You reject the null hypothesis when p < 0.05

**Different types of T-test**
There are three different types of T-tests:
1. Correlated (Paired/ Dependent) T-test: The correlated T test is performed when the samples typically consist of matched pairs of similar units, or when there are cases of repeated measures.
2. Independent equal-variance (pooled) T-test: The equal variance T test is used when the number of samples in each groups is the same, or the variance of the two data sets is similar.
3. Independent unequal-variance T-test: The unequal variance T test is used when the number of samples in each group is different and the variance of the two data sets is also different. This test is also called the **Welch's t-test.** 

![Selection of T-Test](https://wiki.mu-sigma.com/muwiki/images/c/c7/Select_ttest.png)

Reading material:
1. https://www.investopedia.com/terms/t/t-test.asp
2. https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.ttest_ind.html
3. http://blog.minitab.com/blog/adventures-in-statistics-2/understanding-t-tests-t-values-and-t-distributions

In [97]:
#independent equal variance tests

# generate two independent samples
# define sample size:N
N=100
#Gaussian distributed data with mean = 2 and var = 1
data1 = np.random.randn(N) + 2
#Gaussian distributed data with with mean = 0 and var = 1
data2 = np.random.randn(N)
# compare samples
tstat, p = st.ttest_ind(data1, data2)

print("t = " + str(tstat))
print("two tailed p-value = " + str(2*p))
print("one tailed p-value = " + str(p))

t = 15.56673473136405
two tailed p-value = 6.583946221216434e-36
one tailed p-value = 3.291973110608217e-36


Please note: The two-tail P value is twice the one-tail P value (assuming you correctly predicted the direction of the difference).

for **correlated T-test**: try the *ttest_rel* function from the same library


for **independent unequal variance test**: use the equal_var : bool, optional
If True (default), perform a standard independent 2 sample test that assumes equal population variances. If False, perform Welch’s t-test, which does not assume equal population variance

### F Test

An F-test is used to test if the variances of two populations are equal. Just like other hypothesis tests, this test can be a two-tailed test or a one-tailed test. The two-tailed version tests against the alternative that the variances are not equal. The one-tailed version only tests in one direction, that is the variance from the first population is either greater than or less than (but not both) the second population variance. The choice is determined by the problem. For example, if we are testing a new process, we may only be interested in knowing if the new process is less variable than the old process.

In summary, the F-test can be used to answer the following questions:
Do two samples come from populations with equal variancess?
Does a new process, treatment, or test reduce the variability of the current process?

**Basic Assumptions:**
Your population must be **approximately normally distributed (i.e. fit the shape of a bell curve)** in order to use the test. Plus, the samples must be **independent events**. In addition, there are a few important points:

1. The larger variance should always go in the numerator (the top number) to force the test into a right-tailed test. Right-tailed tests are easier to calculate.
2. For two-tailed tests, divide alpha by 2 before finding the right critical value.
3. If you are given standard deviations, they must be squared to get the variances.

The F hypothesis test is defined as:

H0 (null hypothesis):	(σ1)^2=(σ2)^2

Ha (alternative hypothesis):	
(σ1)^2<(σ2)^2	   	for a lower one-tailed test;

(σ1)^2>(σ2)^2	   	for an upper one-tailed test;

(σ1)^2≠(σ2)^2	   	for a two-tailed test;

Test Statistic:	F = (s1)^2/(s2)^2

where (s1)^2 and (s2)^2 and are the sample variances. The more this ratio deviates from 1, the stronger the evidence for unequal population variances.

**How to consume the reults of the F-test?**
You need to compare the computed F-value with the critical F-value obtained from the F-table on the basis of the degrees of freedom. If your calculated value is higher than the table value, you can reject the null hypothesis.

You can also use the associated p-value to accept or reject the hypothesis. If p-value< alpha (generally 0.05), reject the null-hypothesis.

Additional reading material:
1. https://www.statisticshowto.datasciencecentral.com/probability-and-statistics/hypothesis-testing/f-test/ and https://www.statisticshowto.datasciencecentral.com/tables/f-table/ 

In [98]:
from PyAstronomy import pyasl

In [99]:
# generate two independent samples
# define sample size:N
N1=50
N2=20
ch1=900
ch2=100
#Gaussian distributed data with mean = 2 and var = 900
data1 = np.sqrt(ch1) * np.random.randn(N1) + 2
#Gaussian distributed data with with mean = 0 and var = 100
data2 = np.sqrt(ch2) * np.random.randn(N2)

print(data1)
print(data2)

dof1=len(data1)-1
dof2=len(data2)-1

print('dof1:', dof1)
print('dof2:', dof2)

# simple Fisher's classical test
# compare samples
fval = pyasl.ftest(ch1,ch2,dof1, dof2)
print("Test Stats = " + str(fval))

[ 47.70582597  -1.32203058  30.66849347 -17.20159544  -2.89902441
  36.10959867 -67.16032066  -5.98387133  11.31341947  37.11063587
  12.46320061  56.96382437  46.89575867 -26.16749361  22.13872419
  26.73464203  18.03711776  13.40600488  24.55749762  47.23028627
 -17.18334608   1.04089621 -18.00037599  23.44277722  46.7384864
  -6.58805741 -27.41225595 -15.09141931 -40.20518903  25.56353673
  32.270431    12.7877844   39.72983135 -47.52377031  17.02103989
   7.44731445  -7.64639729   7.4872479    0.80478955  22.25139817
 -47.19218759  14.90560505  -8.1719195  -20.44936856 -30.41926798
 -12.84060523  23.80672931  19.49441783  -9.47703128  -0.53069519]
[ -9.22143432   1.27456373 -18.13991654  -4.94164353   5.12786045
   8.77920302  -1.81972474   1.16754898  -4.65653113   3.41834204
   7.75260619   3.70345559 -14.29654719  -8.17986913  11.09569891
   1.50708905  11.87024889  -1.50380106  -1.92848752 -11.27890823]
dof1: 49
dof2: 19
Test Stats = {'F statistic': 5.066666666666666, 'p-value'

package documentation can be found here: https://www.hs.uni-hamburg.de/DE/Ins/Per/Czesla/PyA/PyA/pyaslDoc/aslDoc/statTest.html

### ANOVA

An ANOVA test or Analysis of Variance is a way to find out if survey or experiment results are significant. More specifically, it states whether your sample value closely matches what value you would expect to find in an entire population

There are multiple kinds of ANOVA tests:
1. One-way ANOVA test - A one way ANOVA is used to compare two means from two independent (unrelated) groups using the F-distribution. The null hypothesis for the test is that the two means are equal. Therefore, a significant result means that the two means are unequal.
2. Two Way ANOVA - A Two Way ANOVA is an extension of the One Way ANOVA. With a One Way, you have one independent variable affecting a dependent variable. With a Two Way ANOVA, there are two independents. Use a two way ANOVA when you have one measurement variable (i.e. a quantitative variable) and two nominal variables. In other words, if your experiment has a quantitative outcome and you have two categorical explanatory variables, a two way ANOVA is appropriate.
3. Factorial ANOVA - A factorial ANOVA is an Analysis of Variance test with more than one independent variable, or “factor“. It can also refer to more than one Level of Independent Variable. 

**How do we interpret results?**
ANOVA tests give you an F-value and a corresponding p-value. Like all other tests, you can reject the null hypothesis if p<alpha (0.05)

In [100]:
# One Way ANOVA test

#create data
tillamook = [0.0571, 0.0813, 0.0831, 0.0976, 0.0817, 0.0859, 0.0735,0.0659, 0.0923, 0.0836]
newport = [0.0873, 0.0662, 0.0672, 0.0819, 0.0749, 0.0649, 0.0835,0.0725]
petersburg = [0.0974, 0.1352, 0.0817, 0.1016, 0.0968, 0.1064, 0.105]
magadan = [0.1033, 0.0915, 0.0781, 0.0685, 0.0677, 0.0697, 0.0764,0.0689]
tvarminne = [0.0703, 0.1026, 0.0956, 0.0973, 0.1039, 0.1045]

#conduct the test
fval, pval=st.f_oneway(tillamook, newport, petersburg, magadan, tvarminne)

print("fval = " + str(fval))
print("p-value = " + str(pval))

fval = 7.121019471642447
p-value = 0.0002812242314534544


In [101]:
from statsmodels.formula.api import ols
import statsmodels.api as sm

In [102]:
# Two Way ANOVA Test

#data file
datafile = "ToothGrowth.csv"
df = pd.read_csv(datafile)

# Fits the model with the interaction term
# This will also automatically include the main effects for each factor
formula = 'len ~ C(supp) + C(dose) + C(supp):C(dose)'
model = ols(formula, df).fit()

# Seeing if the overall model is significant
print(f"Overall model F({model.df_model: .0f},{model.df_resid: .0f}) = {model.fvalue: .3f}, p = {model.f_pvalue: .4f}")

print(model.summary())

# Create the ANOVA table
aov_table = sm.stats.anova_lm(model, typ=2)
print(aov_table)

Overall model F( 5, 54) =  41.557, p =  0.0000
                            OLS Regression Results                            
Dep. Variable:                    len   R-squared:                       0.794
Model:                            OLS   Adj. R-squared:                  0.775
Method:                 Least Squares   F-statistic:                     41.56
Date:                Mon, 11 Feb 2019   Prob (F-statistic):           2.50e-17
Time:                        11:55:13   Log-Likelihood:                -159.35
No. Observations:                  60   AIC:                             330.7
Df Residuals:                      54   BIC:                             343.3
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                                   coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------

Let’s break down this ANOVA table. The top 3 rows show the between groups effect of the 3 components (supp, dose and 'supp x dose') which is the overall experimental effect. 

Say, in the first row (supp), the sum of squares for the model (SSM; value 2015.35 in the table) is how much variance is explained by that component in our model. The current model explains a significant amount of variance, F( 5, 54) =  41.557, p < 0.05. The residual row is the unsystematic variation in the data (SSR; also called the unexplained variance; value 712.106 in the table). 

While theseresults give an indication of whether there is a statistically significant effect of some factor on the outcome, it’s as important to know the size of the effect the factor has on the outcome. Statistical significance does not always translate into a large effect in the world.

For the same, we introduce 2 new metrics:
1. eta-squared (η^2): Eta Squared is calculated the same way as R Squared, and has the most equivalent interpretation: out of the total variation in Y, the proportion that can be attributed to a specific X. Eta Squared, however, is used specifically in ANOVA models. Each categorical effect in the model has its own Eta Squared, so you get a specific, intuitive measure of the effect of that variable.

*Please note: Eta Squared has two drawbacks. One is that as you add more variables to the model, the proportion explained by any one variable will automatically decrease. This makes it hard to compare the effect of a single variable in different studies. The second drawback for eta Squared is that it is a biased measure of population variance explained (although it is accurate for the sample). It always overestimates it. This bias gets very small as sample size increases, but for small samples an unbiased effect size measure is Omega Squared.*


2. omega-squared (ω^2): Omega squared (ω2) is a measure of effect size, or the degree of association for a population. It is an estimate of how much variance in the response variables are accounted for by the explanatory variables. Omega squared is widely viewed as a lesser biased alternative to eta-squared, especially when sample sizes are small.




In [103]:
#functions to add omega squared and eta squared
def eta_squared(aov):
    aov['eta_sq'] = 'NaN'
    aov['eta_sq'] = aov[:-1]['sum_sq']/sum(aov['sum_sq'])
    return aov
 
def omega_squared(aov):
    #use residual to calculate mean squared error
    mse = aov['sum_sq'][-1]/aov['df'][-1]
    aov['omega_sq'] = 'NaN'
    aov['omega_sq'] = (aov[:-1]['sum_sq']-(aov[:-1]['df']*mse))/(sum(aov['sum_sq'])+mse)
    return aov

#applying the function to aov_table
eta_squared(aov_table)
omega_squared(aov_table)
print(aov_table)

                      sum_sq    df          F        PR(>F)    eta_sq  \
C(supp)           205.350000   1.0  15.571979  2.311828e-04  0.059484   
C(dose)          2426.434333   2.0  91.999965  4.046291e-18  0.702864   
C(supp):C(dose)   108.319000   2.0   4.106991  2.186027e-02  0.031377   
Residual          712.106000  54.0        NaN           NaN       NaN   

                 omega_sq  
C(supp)          0.055452  
C(dose)          0.692579  
C(supp):C(dose)  0.023647  
Residual              NaN  


ω2 is a better measure of effect size since it’s unbiased in it’s calculation. It takes into account the degrees of freedom, whereas η2 does not. Side note, η2 and R2 are the same thing in the ANOVA framework.

Higher the value of ω2 (or η2), more significant is the variable.

Additional Reading Material:
1. Effect size metrics: https://www.uccs.edu/lbecker/glm_effectsize
2. 2 way ANOVA steps: https://pythonfordatascience.org/anova-2-way-n-way/
3. Package documentation: https://www.statsmodels.org/dev/generated/statsmodels.stats.anova.anova_lm.html and https://www.statsmodels.org/dev/examples/notebooks/generated/interactions_anova.html

### Chi squared test

Chi-square test (or **Pearson’s Chi-Squared Test**) is a nonparametric test used for two specific purpose: 

(a) To test the hypothesis of no association between two or more groups, population or criteria (i.e. to check independence between two variables); 

(b) and to test how likely the observed distribution of data fits with the distribution that is expected (i.e., to test the goodness-of-fit). 

**Please note: It is used to analyze categorical data (e.g. male or female patients, smokers and non-smokers, etc.), it is not meant to analyze parametric or continuous data (e.g., height measured in centimeters or weight measured in kg, etc.).**

An ideal Hypothesis for chi-squared tests:

1.Null hypothesis: Assumes that there is no association between the two variables.

2.Alternative hypothesis: Assumes that there is an association between the two variables.

**In other words, the Chi-Squared test is a statistical hypothesis test that states the null hypothesis as:
two categorical variables are independent in a given population.** 


**Some basic assumptions:**
1. The data are randomly drawn from a population.
2. The sample size is sufficiently large. The application of the Chi-square test to a smaller sample could lead to type II error (i.e. accepting the null hypothesis when it is actually false). There is no expected cut-off for the sample size; however, the minimum sample size varies from 20 to 50
3. The variables under consideration must be mutually exclusive. It means that each variable must only be counted once in a particular category and should not be allowed to appear in other category. In other, words no item shall be counted twice.

**How to interpret the results?**
The variables are considered independent if the observed (the categorically coded data that you have collected) and expected frequencies (frequencies that you would expect to get in each cell of a table by chance alone ) are similar.

1. If Test Statistic >= Critical Value (p-value <= alpha): significant result, reject null hypothesis (H0), dependent.
2. If Test Statistic < Critical Value (p-value > alpha): not significant result, fail to reject null hypothesis (H0), independent

Additional Reading Material:
1. http://www.j-pcs.org/article.asp?issn=2395-5414;year=2015;volume=1;issue=1;spage=69;epage=71;aulast=Rana
2. https://machinelearningmastery.com/chi-squared-test-for-machine-learning/

In [104]:

# chi-squared test with similar proportions
from scipy.stats import chi2_contingency
from scipy.stats import chi2
# contingency table-- summarizes the count of records between two categorical variables
table = [	[10, 20, 30],
			[6,  9,  17]]
print(table)
#The function (chi2_contingency) takes an array as input representing the contingency table for the two categorical variables. 
#It returns the calculated statistic and p-value for interpretation as well as the calculated degrees of freedom and table of expected frequencies.
stat, p, dof, expected = chi2_contingency(table)
print('dof=%d' % dof)
print(expected)
# interpret test-statistic
prob = 0.95 #(assuming  that the variable is independent)
critical = chi2.ppf(prob, dof) #Percent point function 
print('probability=%.3f, critical=%.3f, stat=%.3f' % (prob, critical, stat))
if abs(stat) >= critical:
	print('Dependent (reject H0)')
else:
	print('Independent (fail to reject H0)')
# interpret p-value
alpha = 1.0 - prob
print('significance=%.3f, p=%.3f' % (alpha, p))
if p <= alpha:
	print('Dependent (reject H0)')
else:
	print('Independent (fail to reject H0)')

[[10, 20, 30], [6, 9, 17]]
dof=2
[[10.43478261 18.91304348 30.65217391]
 [ 5.56521739 10.08695652 16.34782609]]
probability=0.950, critical=5.991, stat=0.272
Independent (fail to reject H0)
significance=0.050, p=0.873
Independent (fail to reject H0)


Package doc:
1. https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.chi2_contingency.html